Refactoring unit file (#669)
COVESA/vehicle_signal_specification: docs-gen/content/rule_set/data_entry/data_unit_types.md

## List of supported Data Unit Types

> The VSS syntax does not in itself specify what units that can be used, the unit attribute as declared for signals in *.vspec files is optional and can contain an arbitrary string value.

Reviewer: ```suggestion
The VSS syntax does not in itself specify what units can be used, the unit attribute as declared for signals in *.vspec files is optional and can contain an arbitrary string value.
```

[VSS-Tools](https://github.com/COVESA/vss-tools) however require that all units used are defined.

> Units are defined by including them in an a unit file with syntax as described below.

Reviewer: ```suggestion
Units are defined by including them in a unit file with syntax as described below.
```

> One or more unit files can be specified by the `-u` parameter, and if not given the tools will search for a file `units.yaml`

Reviewer: ```suggestion
One or more unit files can be specified by the `-u` parameter and, if not given, the tools search for a file `units.yaml`
```

in the same directory as the root *.vspec file.

For the VSS standard catalog the VSS-project has defined a set of units that can be used for signals in the VSS standard catalog.

> This list is composed with definitions according to International Units (SI) and few automotive specific units:

Reviewer: ```suggestion
This list is composed of definitions according to International Units (SI) and few automotive-specific units:
```

[Specification](https://www.iso.org/standard/30669.html), [Wikipedia](https://en.wikipedia.org/wiki/International_System_of_Units)

The VSS list of units for the standard catalog exists in [units.yaml](https://github.com/COVESA/vehicle_signal_specification/blob/master/spec/units.yaml).

## Unit file syntax

> Unit files have the syntax defined below:

Reviewer: ```suggestion
Unit files follow the syntax defined below:
```

```
[
 <vss-unit-identifier>: # Typically unit abbreviation, like km/h or mm, but
    definition: <string>
    [unit: <string>] # Full name of unit, optional, if not given assumed to be equal to vss-unit-identifier
    domain: <string> # Domain of the unit.
    [allowed-datatypes] : ['numeric', 'string', uint8', ...]] # Allowed datatypes in VSS standard catalog
    [deprecation: <reason>]
]*
```

The VSS term `domain` corresponds roughly to the term `quantity` as defined in for example ISO 80000.
Typical example are `length`, `mass` and `velocity` that all can be expressed in SI-units.

> Domains do however not from a VSS perspective need to correspond to physical quantities.

Reviewer: ```suggestion
However, from a VSS perspective domains do not need to correspond to physical quantities.
```

> It could be an arbitrary term, but it is in general expected that it is possible to convert between values

Reviewer: ```suggestion
It could be an arbitrary term, but it is generally expected that it is possible to convert between values
```

using different units but defined with the same domain.

Example:

```
m:
  definition: Length measured in meters
  unit: meter
  domain: length
  allowed-datatypes: ['numeric']

mm:
  definition: Length measured in millimeters
  unit: millimeter
  domain: length
  allowed-datatypes: ['numeric']
```

> As `m` and `mm` are defined with the same domain it is expected that you can convert a value from `m`to `mm`.

Reviewer: ```suggestion
As `m` and `mm` are defined with the same domain it is expected that you can convert a value from `m` to `mm`.
```

> The `allowed-datatypes` attributes can be used to specify which VSS datatypes that can contain values of a particular unit.

Reviewer: ```suggestion
The `allowed-datatypes` attributes can be used to specify which VSS datatypes can contain values of a particular unit.
```

For most units the symbolic datatype `numeric` meaning any float or integer datatype can be used.
As an example, both `uint8` and `float` can be used to represent a length value.
If using `uint8` you will have range restrictions, but that might be acceptable for some signals.
For some units more specific datatype restrictions are relevant. Some examples:

> * A date/time expressed in ISO 8061 format can only be represented in string

Reviewer: ```suggestion
* A date/time expressed in ISO 8061 format can only be represented as a string
```

* A UNIX Timestamp signal must be at least 32 bit unsigned to be able to handle date/time after year 2038.

> The `deprecation` keyword can be used to indicate that a specific unit may be removed in the future,

Reviewer: ```suggestion
The `deprecation` keyword can be used to indicate that a specific unit may be removed in the future.
```

> Tooling shall preferably give a warning if a signal use a deprecated unit or the unit used belongs to a deprecated quantity. The reason should preferably list when and why the unit is deprecated, a hypothetical example is given below:

Reviewer: ```suggestion
Tooling shall preferably give a warning if a signal uses a deprecated unit or the unit used belongs to a deprecated quantity. The reason should preferably list when and why the unit is deprecated, a hypothetical example is given below:
```

```

  inch:
    definition: Distance measured in inches
    unit: inch
    domain: distance

>     deprecation: V5.0 replaced with in

Reviewer: ```suggestion
    deprecation: V5.0 replaced by 'in'
```

    <vss-unit-identifier>: # Typically unit abbreviation, like km/h or mm, but
      label: <string> # Replaced with "unit"
      description: <string> # Replaced with "definition"
      domain: <string>
  ]*
```

## Domain file syntax

Defing domains is recommended, but currently optional for backward compatibility reasons.
If tooling supports domain files it can verify that all units provided in unit files
use defined domains.

``

  [
    <vss-quantity-identifier>: # Identifier preferably taken from a standard, like ISO 80000
      definition: <string>
      [remark: <string>] # remark as defined in for example ISO 80000
      [comment: <string>]
      [deprecation: <reason>]
  ]*

```

The VSS list of domains for the standard catalog exists in [domains.yaml](https://github.com/COVESA/vehicle_signal_specification/blob/master/spec/domains.yaml).

## Defining custom units

It is possible to define custom units in a unit file.
Assume for instance you want to have a signal showing remaining range in [furlong](https://en.wikipedia.org/wiki/Furlong).
Then you could add an additional unit `furlong`. No need to specify `unit` or `symbol` as they equals the default (i.e. "furlong").

> As this is unit not commonly used and not described in any standards it might be relevant to describe how it can be converted to other units.

Reviewer: ```suggestion
As this is unit not commonly used and not described in any standards, it might be relevant to describe how it can be converted to other units.
```
